[view-transitions] Pseudo elements can't be captured.

mattwoodrow opened the pull request:
#### 3ba31ee3b3d5e5a87e409978bc11eb8aa0c7f1c5
<pre>
[view-transitions] Pseudo elements can&apos;t be captured.
<a href="https://bugs.webkit.org/show_bug.cgi?id=273166">https://bugs.webkit.org/show_bug.cgi?id=273166</a>
&lt;<a href="https://rdar.apple.com/126964779">rdar://126964779</a>&gt;

Reviewed by Tim Nguyen.

Pseudo elements (like ::backdrop) don&apos;t have an associated Element, so can&apos;t
be found by ViewTransition&apos;s forEachElementInPaintOrder.

Rather than holding an Element in CapturedElement, hold a Styleable instead
which can represent these pseudo elements as well.

Since the normal Styleable class uses an Element reference, introduces a
WeakStyleable object which uses a WeakPtr to an Element and can (maybe) be
converted back into a Styleable when needed.

Changes most of the ViewTransition code to operator on the RenderElement
from the Styleable, instead of the Element.

Adds capturedInViewTransition to Styleable, so that we can look this up
in cases where the renderer hasn&apos;t yet been created.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Element.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/dom/Element.h:
(WebCore::Element::capturedInViewTransition const): Deleted.
(WebCore::Element::setCapturedInViewTransition): Deleted.
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::effectiveViewTransitionName):
(WebCore::captureOverflowRect):
(WebCore::snapshotElementVisualOverflowClippedToViewport):
(WebCore::forEachRendererInPaintOrder):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::clearViewTransition):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStyles):
(WebCore::ViewTransition::viewTransitionNewPseudoForCapturedElement):
(WebCore::forEachElementInPaintOrder): Deleted.
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::initializeStyle):
(WebCore::RenderElement::capturedInViewTransition const): Deleted.
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::rebuildZOrderLists):
(WebCore::RenderLayer::paintList):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::collectViewTransitionNewContentLayers):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::capturedInViewTransition const):
(WebCore::RenderObject::setCapturedInViewTransition):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::capturedInViewTransition const):
* Source/WebCore/style/Styleable.h:
(WebCore::WeakStyleable::operator bool const):
(WebCore::WeakStyleable::operator=):
(WebCore::WeakStyleable::styleable const):

Canonical link: <a href="https://commits.webkit.org/278123@main">https://commits.webkit.org/278123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/421a793ddaf88b9bc229e96a05f890026afb6dbd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49488 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28770 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52526 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52727 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/161 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51792 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34788 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26390 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40399 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51588 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26313 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42636 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21517 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23773 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43819 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7857 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45687 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44325 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54239 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24571 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20754 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47766 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25842 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46783 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10880 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26682 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25565 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -529,8 +526,10 @@ void ViewTransition::activateViewTransition()
-            newElement->setCapturedInViewTransition(true);
+        if (auto newStyleable = capturedElement->newElement.styleable()) {
+            if (newStyleable->renderer())
+                newStyleable->renderer()->setCapturedInViewTransition(true);
```
```suggestion
            if (CheckedPtr renderer = newStyleable->renderer())
                renderer->setCapturedInViewTransition(true);
```

`Styleable::renderer()` is not free, we should cache the result if we can instead of computing it twice

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -614,8 +613,10 @@ void ViewTransition::clearViewTransition()
-            newElement->setCapturedInViewTransition(false);
+        if (auto newStyleable = capturedElement->newElement.styleable()) {
+            if (newStyleable->renderer())
+                newStyleable->renderer()->setCapturedInViewTransition(false);
```
```suggestion
            if (newStyleable->renderer())
                newStyleable->renderer()->setCapturedInViewTransition(false);
```

ditto

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -692,18 +690,18 @@ ExceptionOr<void> ViewTransition::updatePseudoElementStyles()
             // FIXME: Also check fragmented content here.
-            CheckVisibilityOptions visibilityOptions { .contentVisibilityAuto = true };
-            if (!newElement->checkVisibility(visibilityOptions))
+            if (!newStyleable->renderer())
```
> capturedElement’s [new element](https://www.w3.org/TR/css-view-transitions-1/#captured-element-new-element) has a [flat tree](https://drafts.csswg.org/css-scoping-1/#flat-tree) ancestor that [skips its contents](https://www.w3.org/TR/css-contain-2/#skips-its-contents).

This bit also needs to return failure.

https://drafts.csswg.org/css-view-transitions-1/#style-transition-pseudo-elements-algorithm

mattwoodrow replied:
Ah right, good catch.

nt1m commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -50,6 +52,31 @@ enum class ViewTransitionPhase : uint8_t {
     Done
 };
 
+class WeakStyleable {
```
Should this be in Styleable.h or its own file?

mattwoodrow replied:
Possibly! Bit of a stylistic decision, I can move it to Styleable.h.

nt1m commented on Source/WebCore/style/StyleResolver.h:
```diff
@@ -91,11 +91,11 @@ class Resolver : public RefCounted<Resolver>, public CanMakeSingleThreadWeakPtr<
-    ResolvedStyle styleForElement(const Element&, const ResolutionContext&, RuleMatchingBehavior = RuleMatchingBehavior::MatchAllRules);
+    ResolvedStyle styleForElement(Element&, const ResolutionContext&, RuleMatchingBehavior = RuleMatchingBehavior::MatchAllRules);
 
-    void keyframeStylesForAnimation(const Element&, const RenderStyle& elementStyle, const ResolutionContext&, BlendingKeyframes&);
+    void keyframeStylesForAnimation(Element&, const RenderStyle& elementStyle, const ResolutionContext&, BlendingKeyframes&);
 
-    WEBCORE_EXPORT std::optional<ResolvedStyle> styleForPseudoElement(const Element&, const PseudoElementRequest&, const ResolutionContext&);
+    WEBCORE_EXPORT std::optional<ResolvedStyle> styleForPseudoElement(Element&, const PseudoElementRequest&, const ResolutionContext&);
```
Is it as a result of creating a styleable or something else?

mattwoodrow replied:
It is indeed, `Styleable` takes a mutable `Element`.

We only need the subset of `Styleable` that uses it in a const way, but I haven't thought of a reasonable way to abstract that.

nt1m commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -421,38 +427,38 @@ void Adjuster::adjust(RenderStyle& style, const RenderStyle* userAgentAppearance
         adjustDisplayContentsStyle(style);
 
-    if (m_element && (m_element->hasTagName(frameTag) || m_element->hasTagName(framesetTag))) {
+    if (nonPseudoElement() && (nonPseudoElement()->hasTagName(frameTag) || nonPseudoElement()->hasTagName(framesetTag))) {
```
```suggestion
    if (nonPseudoElement() && (nonPseudoElement()->hasTagName(frameTag) || nonPseudoElement()->hasTagName(framesetTag))) {
```

I'm not sure we want to adjust the pseudo-element styles based on the associated element. I think a better pattern is:

```
if (auto element = m_styleable.element) {
    if (element->hasTagName(frameTag) || element->hasTagName(framesetTag))) {
}
```

mattwoodrow replied:
I think that's a behaviour change.

The existing code always passed an Element for non-pseudos, and never passed one for pseudos (if I read all the constructors correctly).

The intention is that `nonPseudoElement` returns a pointer in the same circumstances that `m_element` was non-null.

`m_styleable.element` is never null (it's a reference), so this change would do the tag checks on the originating element of a pseudo, which I think is incorrect.

nt1m replied:
Hmm I see, I wonder if we can name `nonPseudoElement` better like `rendererElement` (as an analogy for `renderer->element()`) or something similar. I'm not a huge fan of `nonPseudoElement`

mattwoodrow replied:
RenderElement::nonPseudoElement is the equivalent function we’d call if we had a renderer.

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -355,45 +346,47 @@ ExceptionOr<void> ViewTransition::captureOldState()
-        auto result = forEachElementInPaintOrder([&](Element& element) -> ExceptionOr<void> {
-            if (auto name = effectiveViewTransitionName(element); !name.isNull()) {
+        auto result = forEachRendererInPaintOrder([&](RenderLayerModelObject& renderer) -> ExceptionOr<void> {
+            if (!Styleable::fromRenderer(renderer) || renderer.isSkippedContent())
```
Can we move the skipped content check in `effectiveViewTransitionName`?

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -407,16 +400,20 @@ ExceptionOr<void> ViewTransition::captureNewState()
-            if (auto name = effectiveViewTransitionName(element); !name.isNull()) {
+        auto result = forEachRendererInPaintOrder([&](RenderLayerModelObject& renderer) -> ExceptionOr<void> {
+            std::optional<const Styleable> styleable = Styleable::fromRenderer(renderer);
+            if (!styleable  || renderer.isSkippedContent())
```
```suggestion
            if (!styleable || renderer.isSkippedContent())
```

nt1m commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -811,7 +811,13 @@ void RenderLayer::rebuildZOrderLists(std::unique_ptr<Vector<RenderLayer*>>& posZ
             if (!posZOrderList)
                 posZOrderList = makeUnique<Vector<RenderLayer*>>();
 
+            RenderLayer* viewTransitionLayer = nullptr;
```
```suggestion
            RenderLayer* viewTransitionLayer;
```

mattwoodrow replied:
A stack pointer won't be default initialised, I think we need this.

nt1m reviewed (commented):
This doesn't capture any pseudo-element yet right? I don't see a place where we set the capturedInViewTransition flag for pseudo-elements to true.

mattwoodrow committed:
[view-transitions] Pseudo elements can't be captured.
https://bugs.webkit.org/show_bug.cgi?id=273166
<rdar://126964779>

Reviewed by Tim Nguyen.

Pseudo elements (like ::backdrop) don't have an associated Element, so can't
be found by ViewTransition's forEachElementInPaintOrder.

Rather than holding an Element in CapturedElement, hold a Styleable instead
which can represent these pseudo elements as well.

Since the normal Styleable class uses an Element reference, introduces a
WeakStyleable object which uses a WeakPtr to an Element and can (maybe) be
converted back into a Styleable when needed.

Changes most of the ViewTransition code to operator on the RenderElement
from the Styleable, instead of the Element.

Adds capturedInViewTransition to Styleable, so that we can look this up
in cases where the renderer hasn't yet been created.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Element.cpp:
(WebCore::operator<<):
* Source/WebCore/dom/Element.h:
(WebCore::Element::capturedInViewTransition const): Deleted.
(WebCore::Element::setCapturedInViewTransition): Deleted.
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::effectiveViewTransitionName):
(WebCore::captureOverflowRect):
(WebCore::snapshotElementVisualOverflowClippedToViewport):
(WebCore::forEachRendererInPaintOrder):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::clearViewTransition):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStyles):
(WebCore::ViewTransition::viewTransitionNewPseudoForCapturedElement):
(WebCore::forEachElementInPaintOrder): Deleted.
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::initializeStyle):
(WebCore::RenderElement::capturedInViewTransition const): Deleted.
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::rebuildZOrderLists):
(WebCore::RenderLayer::paintList):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::collectViewTransitionNewContentLayers):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::capturedInViewTransition const):
(WebCore::RenderObject::setCapturedInViewTransition):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::capturedInViewTransition const):
* Source/WebCore/style/Styleable.h:
(WebCore::WeakStyleable::operator bool const):
(WebCore::WeakStyleable::operator=):
(WebCore::WeakStyleable::styleable const):

Canonical link: https://commits.webkit.org/278123@main
